editor/inserter/index.js

```diff
@@ -1,18 +1,31 @@
+/**
+ * External dependencies
+ */
+import { groupBy } from 'lodash';
+
 const Inserter = () => {
 	const blocks = wp.blocks.getBlocks();
+	const blocksByCategory = groupBy( blocks, 'category' );
+	const categories = wp.blocks.getCategories();
 
 	return (
 		<div className="inserter">
 			<div className="inserter__arrow" />
 			<div className="inserter__content">
-				<div className="inserter__Category-blocks">
-					{ blocks.map( ( { slug, title, icon } ) => (
-						<div key={ slug } className="inserter__block">
-							<span className={ 'dashicons dashicons-' + icon } />
-							{ title }
+				{ categories
```

Comment:
I updated my previous example with a more compelling argument for `Array#reduce` preserving functionally composing the filter / map:

```js
const filterHasBlocks = () => /* ... */;
const mapRenderCategory = () => /* ... */;

categories.reduce( ( memo, category ) => (
	memo.concat( flow( filterHasBlocks, mapRenderCategory )( [ category ] ) )
), [] )
```

_(h/t @mrbobbybryant)_